### PR TITLE
Split up miri into the librustc_mir and bin parts

### LINKED ISSUE #275: The Machine ™
# Things that need to be moved to the machine before the merger

* [ ] ~~array length computation via `Rvalue::Len`~~
* [ ] heap allocations via `Rvalue::NullaryOp(NullOp::Box, _)`
* [ ] `memory::Kind` has `Rust` (heap), `C` (malloc heap) and `Env` (env var emulation) allocations
    * We could add a generic `Machine` variant to `Kind`, which every `Machine` can override
* [ ] ~~branching (`TerminatorKind::SwitchInt`) isn't allowed in CTFE in Rust right now~~

# Things I'm not sure about

There's a lot of stuff that moving to The Machine will make the code much more complex, even though many of these things are basically trivial additions to const eval that imo Rust should have, but couldn't have had with HIR-based const eval.

* [ ] ~~Function pointers and calling them~~ only `const fn` is already a guarantee
* [ ] ~~Some casts which aren't supported by CTFE in Rust right now~~
* [ ] ~~Unsizing~~
* [ ] ~~**Local variables**~~
* [ ] ~~Destructuring~~
* [ ] ~~`Rvalue::Discriminant`~~
* [ ] ~~`unions`~~
* [ ] ~~`enum` creation~~
* [ ] ~~References and dereferencing~~
* [ ] ~~closures and calling them~~
* [ ] unsafe code handcrafting a vtable by moving a pointer to a `const fn` into the appropriate vtable fields and then calling the method via a fat pointer pointing to the field.

Some of these things are really hard to untangle from miri without making the code quite riddiculous. 

Please check boxes on the things that need to be done in the "I'm not sure"-list and discuss the rest below.

### PR DESCRIPTION
**Pull request by oli-obk**
cc @RalfJung this will break your rebasing

I adjusted all the imports so they appear like they would inside `librustc_mir`. For now there's a miri crate that only reexports the `librustc_mir::interpret` contents. This is so that I can write a `Machine` trait which will allow us to rip out everything non CTFE from the `librustc_mir` and place it back into `miri`.

Next step: break all lines longer than 100 chars.

fixes #275

### CODE REVIEW

**Comment by RalfJung** [comment]
How do you feel about accepting some of my validation stuff even though it is currently dead code because rustc does not yet have the validation commands? Would you be fine with me experimenting in-tree (the design of the validation statements is certainly not yet final)? Also see <https://github.com/rust-lang/rust/pull/43403>.
Rebasing is getting tiresome.^^

**Comment by oli-obk** [comment]
I am absolutely fine with it, but how can it be done? The matches on `Statement` will fail, or do you mean just the relevant functions?

Alternatively I can offer running your branch in CI so any miri change has to build a stage 1 rebase your miri over the PR and run the tests. That would keep it out of tree, but ensure that it's the PR author's (my) job to create a new branch with your code rebased over the PR.

Depends on your projected timeline on getting your rustc PR merged. If it's a week or two, let's just get your dead code in here, if it's a few months, the other solution might be better( but for convenience it would also take all the code that compiles into miri)

You might even be able to get all your code merged under a cargo feature. Not sure if cfgs can create match arms

**Comment by RalfJung** [comment]
> but how can it be done? The matches on Statement will fail, or do you mean just the relevant functions?

We could merge only the relevant new bits in memory.rs, or we could also merge the new file validation.rs which contains most of the rest of the new stuff. (The git history involves lots of changes in lvalue.rs, but almost all of that is later moved to validation.rs. I am not sure if you prefer a clean or a true history. ;) Once https://github.com/rust-lang/rust/pull/43403 lands the small patch to hook this all up with the statements could be added. Notice that none of this changes existing behavior; rustc only keeps the new statements around when `-Z mir-emit-validate` is set.

> Depends on your projected timeline on getting your rustc PR merged. If it's a week or two, let's just get your dead code in here, if it's a few months, the other solution might be better( but for convenience it would also take all the code that compiles into miri)

That mostly depends on the compiler team. ;)

**Comment by oli-obk** [comment]
> I am not sure if you prefer a clean or a true history

true is fine by me.

So I suggest you open a PR with your branch, and we merge that before messing further with the structure of miri.

**Comment by oli-obk** [comment]
@eddyb This PR now has a minimal `Machine` trait + its miri and const eval impls.

More methods will follow.

File: src/librustc_mir/interpret/const_eval.rs [hunk 1]
```diff
@@ -76,3 +78,21 @@ pub fn eval_body_as_integer<'a, 'tcx>(
         _ => return Err(EvalError::NeedsRfc("evaluating anything other than isize/usize during typeck".to_string())),
     })
 }
+
+struct Evaluator;
```
Comment (eddyb) [inline src/librustc_mir/interpret/const_eval.rs]: I'd put `Const` in its name. Or `CompileTime`.

Comment (oli-obk) [inline src/librustc_mir/interpret/const_eval.rs]: OK. Otherwise this is along the lines you had in mind?

Comment (eddyb) [inline src/librustc_mir/interpret/const_eval.rs]: Look good to me!

File: miri/src/lib.rs [hunk 2]
```diff
@@ -1,3 +1,789 @@
...
+        // as if the call just completed and it's returning to the
+        // current frame.
+        self.dump_local(dest);
+        self.goto_block(dest_block);
+        return Ok(());
+    }
+}
+
+//struct Memory<'b, 'a: 'b, 'tcx: 'a>(&'b rustc_miri::interpret::Memory<'a, 'tcx, Evaluator>);
+struct MemoryMut<'b, 'a: 'b, 'tcx: 'a>(&'b mut rustc_miri::interpret::Memory<'a, 'tcx, Evaluator>);
```
Comment (RalfJung) [inline miri/src/lib.rs]: What is the purpose of this newtype?

Comment (RalfJung) [inline miri/src/lib.rs]: (And same for the other newtype above.)

**Comment by RalfJung** [comment]
I *think* I like the overall design, but I am not sure because I cannot see any structure in `miri/src/lib.rs`. ;) That file is rather... messy. I haven't even found the place yet where it says that this particular `Machine` impl is supposed to be used.

**Comment by RalfJung** [comment]
So the folder structure is such that we have `src/librustc_mir` and `miri/src`? That looks a little funny.
It'd be great if there could be some folder that has a `Cargo.toml` and that contains everything. For example, what about putting miri's `Cargo.toml` into the root?

**Comment by RalfJung** [comment]
I was thinking of something like this: https://github.com/RalfJung/miri/commit/587be4f6ab39b8013d438b6ed0d1f5c3eef4230c

File: miri/src/lib.rs (hunk 2 above)
Comment (oli-obk) [inline miri/src/lib.rs]: It allows writing methods for EvalContext. If we wrote an extension trait, we'd have to duplicate every single method on the trait and the impl.

**Comment by oli-obk** [comment]
The librustc_Mir folder structure is mirroring the rustc repository in order to simplify the merger into rustc. I'd rather not change the folder structure, but I can place a Cargo.toml into the root instead of in the miri folder

File: miri/src/lib.rs (hunk 2 above)
Comment (RalfJung) [inline miri/src/lib.rs]: I see, makes sense. I find it a little funny that this ends up passing around `&mut &mut ...`, but I guess that's needed to get automatic reborrowing.

For consistency, shouldn't this one be called `Memory` (just like the other one is called `EvalContext`)?

**Comment by RalfJung** [comment]
> The librustc_Mir folder structure is mirroring the rustc repository in order to simplify the merger into rustc. 

Oh, I see. So you plan to use git-filter or so to then extract just this subfolder? That would still kill most of the history though.

**Comment by RalfJung** [comment]
So what about <https://github.com/RalfJung/miri/commit/fa9a15fcf792c94c81da879b7025750d1d491826>? The idea is also to have the folder structure as much as possible already like how we'd like to have it after the merge -- the only exception being that the miri source is in miri, not src. It could be in src, but that'd be rather confusing with librustc_mir also being there.

**Comment by oli-obk** [comment]
Sgtm. I already have this rebased over master locally. I'll apply the move from your commit tomorrow and push it here.

> but I am not sure because I cannot see any structure in miri/src/lib.rs. ;) That file is rather... messy. I haven't even found the place yet where it says that this particular Machine impl is supposed to be used.

I have split the file into clearer components. Will also include it in this PR

**Comment by RalfJung** [comment]
> Sgtm. I already have this rebased over master locally. I'll apply the move from your commit tomorrow and push it here.

Alternatively, push this somewhere and I can do the move again.

**Comment by oli-obk** [comment]
Ok, this is ready for review now :)

File: miri/src/lib.rs (hunk 2 above)
Comment (oli-obk) [inline miri/src/lib.rs]: Ok... I gave up on the newtype idea. It gets messy fast. We're back to extension traits. While those create some duplication of the function headers, they are so much more convenient to use.

**Comment by oli-obk** [comment]
Down to 7k lines of code in `interpret`. Slowly moving towards a manageable size for the rustc merger.

**Comment by oli-obk** [comment]
Ok. I think we have reached everything for the merger.

File: miri/lib.rs [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+    fn finish(&mut self) -> EvalResult<'tcx>;
+}
+
+impl<'a, 'tcx> EvalContextExt<'tcx> for EvalContext<'a, 'tcx, Evaluator> {
+    fn finish(&mut self) -> EvalResult<'tcx> {
+        let mut dtor = self.memory.fetch_tls_dtor(None)?;
+        // FIXME: replace loop by some structure that works with stepping
+        while let Some((instance, ptr, key)) = dtor {
+            trace!("Running TLS dtor {:?} on {:?}", instance, ptr);
+            // TODO: Potentially, this has to support all the other possible instances? See eval_fn_call in terminator/mod.rs
```
Comment (RalfJung) [inline miri/lib.rs]: This file name reference needs updating, I presume?

File: miri/lib.rs [hunk 4]
```diff
@@ -0,0 +1,342 @@
...
+    }
+}
+
+trait MemoryExt<'tcx> {
+    fn create_tls_key(&mut self, dtor: Option<ty::Instance<'tcx>>) -> TlsKey;
+    fn delete_tls_key(&mut self, key: TlsKey) -> EvalResult<'tcx>;
+    fn load_tls(&mut self, key: TlsKey) -> EvalResult<'tcx, Pointer>;
+    fn store_tls(&mut self, key: TlsKey, new_data: Pointer) -> EvalResult<'tcx>;
+    fn fetch_tls_dtor(&mut self, key: Option<TlsKey>) -> EvalResult<'tcx, Option<(ty::Instance<'tcx>, Pointer, TlsKey)>>;
+}
```
Comment (RalfJung) [inline miri/lib.rs]: IMHO the TLS stuff should get its own file.

File: miri/lib.rs [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+    }
+
+    fn box_alloc<'a>(
+        ecx: &mut EvalContext<'a, 'tcx, Self>,
+        ty: ty::Ty<'tcx>,
+    ) -> EvalResult<'tcx, PrimVal> {
+        let size = ecx.type_size(ty)?.expect("box only works with sized types");
+        let align = ecx.type_align(ty)?;
+        if size == 0 {
+            Ok(PrimVal::Bytes(align.into()))
```
Comment (RalfJung) [inline miri/lib.rs]: What happened with the FIXME to call the lang item instead?

File: miri/missing_fns.rs [hunk 6]
```diff
@@ -0,0 +1,518 @@
...
+        let args_res: EvalResult<Vec<Value>> = arg_operands.iter()
+            .map(|arg| self.eval_operand(arg))
+            .collect();
+        let args = args_res?;
+
+        let usize = self.tcx.types.usize;
+    
+        match &path[..] {
+            // Allocators are magic.  They have no MIR, even when the rest of libstd does.
+            "alloc::heap::::__rust_alloc" => {
```
Comment (RalfJung) [inline miri/missing_fns.rs]: This is an exact duplicate of some code further up in `call_fn`.

File: src/librustc_mir/interpret/eval_context.rs [hunk 7]
```diff
@@ -16,28 +16,34 @@ use syntax::codemap::{self, DUMMY_SP, Span};
...
+    pub tcx: TyCtxt<'a, 'tcx, 'tcx>,
 
     /// The virtual memory system.
-    pub(crate) memory: Memory<'a, 'tcx>,
+    pub memory: Memory<'a, 'tcx, M>,
 
     #[allow(dead_code)]
     // FIXME(@RalfJung): validation branch
     /// Lvalues that were suspended by the validation subsystem, and will be recovered later
     pub(crate) suspended: HashMap<DynamicLifetime, Vec<ValidationQuery<'tcx>>>,
```
Comment (RalfJung) [inline src/librustc_mir/interpret/eval_context.rs]: So the plan is to keep the validation stuff for CTFE, or to factor it out later?

File: src/librustc_mir/interpret/operator.rs [hunk 8]
```diff
@@ -266,6 +203,9 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
             (Div, F64) => f64_arithmetic!(/, l, r),
             (Rem, F64) => f64_arithmetic!(%, l, r),
 
+            (Eq, _) => PrimVal::from_bool(l == r),
+            (Ne, _) => PrimVal::from_bool(l != r),
```
Comment (RalfJung) [inline src/librustc_mir/interpret/operator.rs]: Why is this needed now?

File: src/librustc_mir/interpret/const_eval.rs [hunk 9]
```diff
@@ -70,6 +79,122 @@ pub fn eval_body_as_integer<'a, 'tcx>(
...
+    }
+
+    fn ptr_op<'a>(
+        _ecx: &EvalContext<'a, 'tcx, Self>,
+        _bin_op: mir::BinOp,
+        _left: PrimVal,
+        _left_ty: Ty<'tcx>,
+        _right: PrimVal,
+        _right_ty: Ty<'tcx>,
+    ) -> EvalResult<'tcx, Option<(PrimVal, bool)>> {
```
Comment (RalfJung) [inline src/librustc_mir/interpret/const_eval.rs]: If I read the code in operator.rs correctly, this function is always called, not just on pointers (and after all, integers could be pointers). So isn't this failing too much?

I think what I'd expect is a test whether _left or _right is a Ptr value. If both are Bytes, we should return Ok(None) rather than Err.

File: src/librustc_mir/interpret/step.rs [hunk 10]
```diff
@@ -150,83 +154,89 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-        } else {
-            return;
-        }
-        if let Err(e) = f(self) {
-            *self.new_constants = Err(e);
+// WARNING: make sure that any methods implemented on this type don't ever access ecx.stack
+// this includes any method that might access the stack
+// basically don't call anything other than `load_mir`, `alloc_ptr`, `push_stack_frame`
+// The reason for this is, that `push_stack_frame` modifies the stack out of obvious reasons
+struct ConstantExtractor<'a, 'b: 'a, 'tcx: 'b, M: Machine<'tcx> + 'a> {
```
Comment (RalfJung) [inline src/librustc_mir/interpret/step.rs]: I never really understood this comment. Why does the fact that `push_stack_frame` push a stack frame entail that all methods on this type (?!?) should not access the stack? What exactly goes wrong if they do?

(This doesn't have to be solved by this PR.)

File: Cargo.toml [hunk 11]
```diff
@@ -35,4 +35,4 @@ compiletest_rs = "0.2.6"
 
 [workspace]
 members = ["src/librustc_mir"]
-exclude = ["xargo"]
+exclude = ["xargo", "cargo-miri-test"]
```
Comment (RalfJung) [inline Cargo.toml]: The `members` line is probably redundant; everything below this folder is in the workspace automatically.

File: miri/missing_fns.rs (hunk 6 above)
Comment (oli-obk) [inline miri/missing_fns.rs]: whoops, copy paste

File: miri/lib.rs (hunk 5 above)
Comment (oli-obk) [inline miri/lib.rs]: got lost, readded

File: src/librustc_mir/interpret/eval_context.rs (hunk 7 above)
Comment (oli-obk) [inline src/librustc_mir/interpret/eval_context.rs]: I'd leave it in, I don't think we want to allow UB in CTFE ;)

File: src/librustc_mir/interpret/operator.rs (hunk 8 above)
Comment (oli-obk) [inline src/librustc_mir/interpret/operator.rs]: Related to your other comment, `ptr_op` doesn't handle integer ops anymore.

File: src/librustc_mir/interpret/const_eval.rs (hunk 9 above)
Comment (oli-obk) [inline src/librustc_mir/interpret/const_eval.rs]: I decided to change the API that either `left` or `right` must be a `Ptr`.

File: src/librustc_mir/interpret/step.rs (hunk 10 above)
Comment (oli-obk) [inline src/librustc_mir/interpret/step.rs]: Well... the issue is with functions like `type_size`, which access `frame().instance.substs`. If you keep expecting such a function to do something useful, you're gonna get suprised.

File: src/librustc_mir/interpret/const_eval.rs (hunk 9 above)
Comment (oli-obk) [inline src/librustc_mir/interpret/const_eval.rs]: I think my reasoning was that in the future you'll want integers to go through the `ptr_op` API to properly handle the intptr model

Comment (oli-obk) [inline src/librustc_mir/interpret/const_eval.rs]: Oh... we can't do this. libstd calls e.g. `Offset` on two `Bytes` for zsts. There's more ops that would fail.... And now I'm just realizing I misunderstood your original meaning. Fixing properly

File: miri/lib.rs (hunk 4 above)
Comment (oli-obk) [inline miri/lib.rs]: it has its own file now

File: src/librustc_mir/interpret/operator.rs (hunk 8 above)
Comment (RalfJung) [inline src/librustc_mir/interpret/operator.rs]: I see. ptr_op still pretty much has to support them (well, at least in its type), but I guess for the purpose of CTFE it makes sense to split this.

File: miri/lib.rs [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+            self.write_ptr(dest, ptr, ty)?;
+
+            // step until out of stackframes
+            while self.step()? {}
+
+            dtor = match self.memory.fetch_tls_dtor(Some(key))? {
+                dtor @ Some(_) => dtor,
+                None => self.memory.fetch_tls_dtor(None)?,
+            };
+        }
```
Comment (RalfJung) [inline miri/lib.rs]: This loop is TLS code, should also move to the TLS file.

File: src/librustc_mir/interpret/operator.rs [hunk 13]
```diff
@@ -142,82 +145,16 @@ impl<'a, 'tcx> EvalContext<'a, 'tcx> {
...
-                }
-                Add | BitAnd
-                if left_kind == right_kind && (left_kind == usize || left_kind == isize)
-                && left.is_bytes() && right.is_ptr() => {
-                    // This is a commutative operation, just swap the operands
-                    return self.ptr_int_arithmetic(bin_op, right.to_ptr()?, left.to_bytes()? as i128, left_kind == isize);
-                }
-                _ => {}
+            if let Some(handled) = M::ptr_op(self, bin_op, left, left_ty, right, right_ty)? {
+                return Ok(handled);
```
Comment (RalfJung) [inline src/librustc_mir/interpret/operator.rs]: The name doesn't really fit right now, does it? This is called for all int and/or ptr ops after all.

I went through the same trouble when I extended miri to handle more ptr ops while trying to keep the code clean. Separating ptr from int really doesn't make too much sense, except for CTFE.

Maybe a better alternative would be to call it only if either operand is a pointer *or* the operation is `Offset` (this being the only primitive operation that takes arguments of pointer type)? The function should then probably not return an `Option`, *but* for intptrcast it would need to have a chance to mutate the values (i.e., turning them into integers). Not sure though. (Note that, as far as I can tell, the current API would require code duplication for intptrcast -- the code below handling integer operations could not be reused.)

Alternatively, the comment documenting `ptr_op` in `trait Machine` should say that this is always called, not just when pointers are involved.

Comment (oli-obk) [inline src/librustc_mir/interpret/operator.rs]: > Note that, as far as I can tell, the current API would require code duplication for intptrcast -- the code below handling integer operations could not be reused.

I'd assume that `PrimVal::to_bytes` would take a `Memory` argument and do the logic in there.

I renamed the method to `try_ptr_op` and changed the documentation to reflect the changes.

Comment (RalfJung) [inline src/librustc_mir/interpret/operator.rs]: > I'd assume that PrimVal::to_bytes would take a Memory argument and do the logic in there.

Hm. Thinking about it, to model intptrcast faithfully we probably want to eagerly do the allocation when the cast happens, or when the transmute happens -- arithmetic operations should never have to do that, as that would mean they have side-effects and optimizations become horrible.

Anyway, we can discuss that in the intptrcast PR, whenever that happens.
